### ISSUE TITLE
Doesn't understand switch `-frandom-seed=libbacktrace.la`

### ISSUE THREAD

**Issue by raphaelcohn** [midipix-project/slibtool#3]
Whilst building musl-cross-make with slibtool (and after a lot of edits to the various gcc, etc, sources, which hard code some libtool assumptions and don't always pass -L where they should), slibtool breaks just before ar linking (confusing saying it is going to run ar, but actually breaking before it) with this line:-
`dlibtool --tag=CC   --mode=link cc -O3 -g0 -funwind-tables -frandom-seed=libbacktrace.la -W -Wall -Wwrite-strings -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition -Wmissing-format-attribute -Wcast-qual  -g -O2  -static-libstdc++ -static-libgcc  -o libbacktrace.la  atomic.lo dwarf.lo fileline.lo posix.lo print.lo sort.lo state.lo backtrace.lo simple.lo elf.lo mmapio.lo mmap.lo`

which is converted to

`dlibtool --tag=CC --mode=link cc -O3 -g0 -funwind-tables -frandom-seed=libbacktrace.la -W -Wall -Wwrite-strings -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition -Wmissing-format-attribute -Wcast-qual -g -O2 -static-libstdc++ -static-libgcc -o libbacktrace.la atomic.lo dwarf.lo fileline.lo posix.lo print.lo sort.lo state.lo backtrace.lo simple.lo elf.lo mmapio.lo mmap.lo`

Failure happens in `slbt_exec_link_create_dep_file`, around the code that does `stat(depfile,&st)` with a perror of `No such file or directory`. The value of `depfile` here is `.libs/-frandom-seed=libbacktrace.a.slibtool.deps`.

I'm not in a position to produce a patch or PR - my C is rather rusty, and I haven't yet discovered why this option slips in.

**Comment by raphaelcohn** [comment]
I can confirm that `depfile` is correct before entering the `for` loop in this code (ie at `if (!(ectx->fdeps = fopen(depfile,"w")))`). So I think a test before `} else if ((popt = strrchr(*parg,'.')) && !strcmp(popt,".la")) {` is needed, perhaps checking for `-` or `-f`.

**Comment by midipix** [comment]
About a month ago I was able to successfully use slibtool in conjunction with musl-cross-make, and as far as I remember I only had to make one small change in mcm for that to work. I'd be happy to give this a new try, as mcm has changed a lot in the meantime.

Just to confirm, I assume you are using the latest version of mcm? If so, how are you invoking make(1)? Anything else I should keep in mind in order to have an environment similar to yours?

**Comment by raphaelcohn** [comment]
My environment is quite likely to be rather different - it's somewhat
experimental. First off, I'm using Alpine Linux 3.4 as a base system. I'm
then artificially restricting the PATH, so I can work out the minimal
dependencies needed. I build slibtool, ccache and gnu make (the last using
a custom shell script so as to avoid the need to have make). I've altered
the configure scripts in the various gcc tools to rip-out hardcoded usage
of libtool - they ignore slibtool mostly (viz: libtool: lines in the stderr
output). Not doing this makes for a successful build, but it isn't really
using slibtool as far as I can tell. To rip out libtool, I do something
like this:-

```bash
libertine_public_autoconf_fixLibToolPostAutoreconf()
{
    local folderPath="$1"

    find "$folderPath" -type f -name 'ltmain.sh' | while IFS= read -r
filePath
    do
        rm "$filePath"
    done

    # Get rid of fscking hardcoded libtool
    find "$folderPath" -type f -name 'configure' | while IFS= read -r
filePath
    do
        # Get rid of ltmain.sh nonsense
        sed -i -e 's/^LIBTOOL=.*$/LIBTOOL=dlibtool/g' -e
's/^LIBTOOL_DEPS=.*$/LIBTOOL_DEPS=/g' -e
's/^ltmain=.*$/ltmain=NO_SUCH_LTMAIN/g' "$filePath"

        # Hack into the built-in ltmain.sh / libtool embedded in configure,
FFS and make it re-route to dlibtool-static
        sed -i -e 's/^# Generated automatically/exec dlibtool "$@"\n#
Generated automatically/g' "$filePath"
    done

    libertine_public_autoconf_removeCruftThatTriesToReconfigure
"$folderPath"
}

libertine_public_autoconf_removeCruftThatTriesToReconfigure()
{
    local folderPath="$1"

    # Inspired by last answer at
https://stackoverflow.com/questions/33278928/how-to-overcome-aclocal-1-15-is-missing-on-your-system-warning-when-compilin
    # Get rid of autoconf cruft that tries to reconfigure
    find "$folderPath" -type f -name 'Makefile.in' | while IFS= read -r
filePath
    do
        sed -i \
            -e 's/ACLOCAL =.*$/ACLOCAL = true/g' \
            -e 's/AUTOCONF =.*$/AUTOCONF = true/g' \
            -e 's/AUTOHEADER =.*$/AUTOHEADER = true/g' \
            -e 's/AUTOMAKE =.*$/AUTOMAKE = true/g' \
            "$filePath"
    done
}
```

I then find that mpc / mpfr won't like to gmp. This is probably because
slibtool does RPATH handling differently (does it not embed them? Nothing
wrong with that, indeed, I loathe RPATH, but it seems the gcc build process
relies on rpaths?) I then apply a fix like this to mpc / mpfr (and a
similar one to isl):-
`sed -i \
                    -e 's/lib'$sourcePackageName'_la_LIBADD =
/lib'$sourcePackageName'_la_LIBADD = @LDFLAGS@ /g' \
                    <source>/src/Makefile.in`

Otherwise a `-L` the the folder with `.libs/gmp` is missing. The
`$(LDFLAGS)` variable itself has `-L` stripped out of it for some reason
post Makefile creation (eg during build by something), hence `@LDFLAGS`

To do this I use mcm with unpacked sources.

(Yes, there's room for improvement above).

Note that I'm not actually running `autoreconf` despite the function name -
that's to come (I want to be able to build from git, not tarballs, so a
downstream user can fork and track changes and thereby audit change).

Other stuff of note. In `config.mak`:-
```
TARGET = x86_64-BUILD-linux-musl
SOURCES =
/home/raph/libertine/var/tmp/packages/build_musl_cross_make_build/copy/sources
OUTPUT =
/home/raph/libertine/var/tmp/packages/build_musl_cross_make_build/build-sysroot/usr

BINUTILS_VER = 2.25.1
GCC_VER = 5.3.0
MUSL_VER = 1.1.15
GMP_VER = 6.1.0
MPC_VER = 1.0.3
MPFR_VER = 3.1.4
LINUX_VER = 4.4.10
ISL_VER = 0.15
# $(SOURCES):
#     :
#
# $(SOURCES)/config.sub: | $(SOURCES)
#     :
#     #cd $(SOURCES) && sha1sum -c $(PWD)/hashes/$(notdir
$@).$(CONFIG_SUB_REV).sha1
#
# $(SOURCES)/%: hashes/%.sha1 | $(SOURCES)
#     :
#     #cd$(SOURCES) && sha1sum -c $(PWD)/hashes/$(notdir $@).sha1

GCC_CONFIG += --disable-multilib

# Fix makeinfo issues
COMMON_CONFIG += MAKEINFO=false
export AM_MAKEFLAGS = INFO_DEPS=

# Libtool hacks
export LIBTOOL = dlibtool-static

COMMON_CONFIG += --disable-nls
COMMON_CONFIG += --with-debug-prefix-map=$(PWD)=
COMMON_CONFIG += --sysconfdir=/etc --sharedstatedir=/com
--localstatedir=/var
COMMON_CONFIG += --disable-silent-rules
```

I also tweak timestamps on `.l` and `.y` files (eg in binutils) to make
sure they're definitively in the past.

A place to start is probably to verify that my assumption about gcc et al
defaulting to using their internal libtool implementations (removing
ltmain.sh is not enough; it recreates it via configure). I did try using
MAKE="make LIBTOOL=dlibtool" but that causes other problems, it seems.


Raphael Cohn
Chief Architect, StormMQ
Co-Chair, OASIS MQTT Standard for the Internet of Things
Advisor, UK Government Technical Standards Board
https://uk.linkedin.com/in/raphcohn
+44 7590 675 756

UK Office:
Hamblethorpe Farm, Crag Lane, Bradley BD20 9DB, North Yorkshire, United
Kingdom
+44 845 3712 567

Registered office:
16 Anchor Street, Chelmsford, Essex, CM2 0JY, United Kingdom
StormMQ IT Limited is Registered in England and Wales under Company Number
07175657

**Comment by midipix** [comment]
First thing first, let's eliminate some trivial pitfalls... do you 

export MAKE="make LIBTOOL=dlibtool"

(dlibtool instead of slibtool for visible confirmation that it's indeed being used).

There's a mention of the above export in the README file, and also of when you should be using dlibtool-static instead. If I remember correctly, when building musl-cross-make I added the special (dlibtool-static) export to the binutils build step.

For what it's worth, my successful build of mcm (on alpine, btw.) did no involve any patches of the underlying sources (gmp,mpfr,mpc,binutils,gcc) whatsoever. One major objective of slibtool is to not require such patches regardless of how incorrect or assumptive the packages being built are. So in order to solve the real problem, I'd suggest taking a step back and reverting to unmodified source.

PS. going through the various steps might be easier on irc:-)

**Comment by raphaelcohn** [comment]
Thanks for the quick reply. Doing `export MAKE="make LIBTOOL=dlibtool"`
previously (ie on Monday) broke the build, but that might have been down to
using ccache at the same time. I've just started using `dlibtool-static`,
and that's moved things on in leaps and bounds. My current issue now is
right down in the bowels after `xgcc` has been built, when creating
libsupc++. xgcc -dumpmachine produces the build host, which slibtool then
wants to prefix ar with (also fine). However, for some reason, it's not
found that prefixed version of ar on the PATH. That's unlikely to be a
slibtool issue. I'll see if I can sort that, and, if not, I'm happy to
revert back to unmodified source.

I don't have access to IRC. Let me try a few things and update you in a few
hours. Thanks again for the helpful response. It's much appreciated.

Raphael Cohn
Chief Architect, StormMQ
Co-Chair, OASIS MQTT Standard for the Internet of Things
Advisor, UK Government Technical Standards Board
https://uk.linkedin.com/in/raphcohn
+44 7590 675 756

UK Office:
Hamblethorpe Farm, Crag Lane, Bradley BD20 9DB, North Yorkshire, United
Kingdom
+44 845 3712 567

Registered office:
16 Anchor Street, Chelmsford, Essex, CM2 0JY, United Kingdom
StormMQ IT Limited is Registered in England and Wales under Company Number
07175657

**Comment by midipix** [comment]
Glad to help and see that this is moving forward. The $HOST-ar tool should certainly be there after a proper cross-compilation of binutils, so I'd suggest you `make install` from the binutils build directory and ensure that is the case.

Regarding irc: hexchat can easily be installed on alpine (apk add hexchat, I believe), or otherwise I'd just use kiwiirc.com in a browser. Simply choose a nickname and set the channel to #midipix (note that irc.freenode.net is the site's default server).

**Comment by raphaelcohn** [comment]
Yippee! It all builds. Thank you so much. I'm going to go over my changes
again, and add them back in one at a time to verify they are appropriate.

**Comment by midipix** [comment]
Excellent, glad it worked out. My recommendation would be to try to build against the unmodified package sources nonetheless, as maintaining your own changes is going to be too much of a headache as time goes by:-) You might have to add a single export into mcm's makefile, otherwise with slibtool-static for everything that possibly won't be necessary.

**Comment by raphaelcohn** [comment]
Right, I've managed to go through everything again. The original defect is definitely there; it's not a gcc or musl-cross-make thing. It's because the code `} else if ((popt = strrchr(*parg,'.')) && !strcmp(popt,".la")) {` thinks the option `-frandom-seed=libbacktrace.la` is actually a `.la` file, when it is actually a weird old switch to gcc to generate a random seed; for want of a better seed string, the developer of `libbacktrace` figured using the name of the library was a clever way to do this. I'm probably getting this because I'm not using dist tarball sources, so I'm at the mercy of autoconf. For what it's worth, we have to work with what we've got. I completely agree maintaining private changes is a headache. It's one of the reasons I like stripped down linux systems - less to know and less to patch every single day...

The fix is to add something like

    else if (starts_with("-f", *parg))
    {
    	// Ignore things like -frandom-seed=libbacktrace.la, which breaks the next line
    }

where `starts_with` might be

    static bool startsWith(const char *pre, const char *str)
    {
         return strncmp(pre, str, strlen(pre)) == 0;
    }

PS I'm looking forward to giving midipix a go in the near future; it's a brilliant idea. Just need to find some nice clear time...
